File: src/test/java/io/codeka/gaia/config/security/oauth2/OAuth2SuccessHandlerTest.java
package io.codeka.gaia.config.security.oauth2;

import io.codeka.gaia.registries.RegistryOAuth2Provider;
import io.codeka.gaia.teams.OAuth2User;
import io.codeka.gaia.teams.User;
import io.codeka.gaia.teams.repository.UserRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
import org.springframework.security.oauth2.client.registration.ClientRegistration;
import org.springframework.security.oauth2.core.AuthorizationGrantType;
import org.springframework.security.oauth2.core.user.DefaultOAuth2User;
import org.springframework.security.web.savedrequest.DefaultSavedRequest;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class OAuth2SuccessHandlerTest {

    @Mock
    UserRepository userRepository;

    @Mock
    OAuth2AuthorizedClientService oAuth2AuthorizedClientService;

    @Mock
    HttpServletRequest request;

    @Mock
    HttpServletResponse response;

    @Mock
    OAuth2AuthenticationToken authentication;

    @Mock
    HttpSession httpSession;

    private OAuth2SuccessHandler oAuth2SuccessHandler;
    private List<RegistryOAuth2Provider> registryOAuth2Providers;

    @BeforeEach
    void setup() {
        registryOAuth2Providers = new ArrayList<>();
        oAuth2SuccessHandler = new OAuth2SuccessHandler(userRepository, registryOAuth2Providers, oAuth2AuthorizedClientService);

        when(authentication.getName()).thenReturn("django");
        when(request.getSession()).thenReturn(httpSession);
    }

    @Test
    void onAuthenticationSuccess_shouldCreateUser_whenNotExists() throws IOException {
        // when
        when(userRepository.findById(anyString())).thenReturn(Optional.empty());
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication);

        // then
        var captor = ArgumentCaptor.forClass(User.class);
        verify(userRepository).save(captor.capture());
        assertThat(captor.getValue()).isNotNull()
                .hasFieldOrPropertyWithValue("username", "django")
                .hasFieldOrPropertyWithValue("team", null);
    }

    @Test
    void onAuthenticationSuccess_shouldUpdateUser_whenExists() throws IOException {
        // given
        var user = new User("calvin", null);

        // when
        when(userRepository.findById(anyString())).thenReturn(Optional.of(user));
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication);

        // then
        verify(userRepository).save(user);
    }

    @Test
    void onAuthenticationSuccess_shouldFillOAuth2User() throws IOException {
        // given
        var user = new User("frankie", null);
        var oauth2User = new OAuth2User("tarantino", "unchained", null);
        var client = mock(OAuth2AuthorizedClient.class);
        var registration = ClientRegistration
                .withRegistrationId("test_registration_id")
                .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
                .clientId("test_client_id")
                .redirectUriTemplate("test_uri_template")
                .authorizationUri("test_authorization_uri")
                .tokenUri("test_token_uri")
                .build();
        var principal = mock(DefaultOAuth2User.class);
        var oauth2Provider = mock(RegistryOAuth2Provider.class);
        registryOAuth2Providers.add(oauth2Provider);

        // when
        when(userRepository.findById(anyString())).thenReturn(Optional.of(user));
        when(oAuth2AuthorizedClientService.loadAuthorizedClient(any(), anyString())).thenReturn(client);
        when(oauth2Provider.isAssignableFor(anyString())).thenReturn(true);
        when(oauth2Provider.getOAuth2User(principal, client)).thenReturn(oauth2User);
        when(client.getClientRegistration()).thenReturn(registration);
        when(authentication.getPrincipal()).thenReturn(principal);
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication);

        // then
        assertThat(user).hasFieldOrPropertyWithValue("oAuth2User", oauth2User);
    }

    @Test
    void onAuthenticationSuccess_shouldRedirectToHomePage() throws IOException {
        // when
        when(userRepository.findById(anyString())).thenReturn(Optional.empty());
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication);

        // then
        verify(response).sendRedirect("/");
    }

    @Test
    void onAuthenticationSuccess_shouldRedirectToAskedPage_whenSpecified() throws IOException {
        // given
        var savedRequest = mock(DefaultSavedRequest.class);

        // when
        when(userRepository.findById(anyString())).thenReturn(Optional.empty());
        when(httpSession.getAttribute("SPRING_SECURITY_SAVED_REQUEST")).thenReturn(savedRequest);
        when(savedRequest.getRequestURI()).thenReturn("/test_url");
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication);

        // then
        verify(response).sendRedirect("/test_url");
    }
}

File: src/test/java/io/gaia_app/config/security/oauth2/OAuth2SuccessHandlerTest.kt
package io.gaia_app.config.security.oauth2

import io.gaia_app.registries.RegistryOAuth2Provider
import io.gaia_app.organizations.OAuth2User
import io.gaia_app.organizations.User
import io.gaia_app.organizations.repository.UserRepository
import io.gaia_app.test.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor.forClass
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.security.oauth2.client.OAuth2AuthorizedClient
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import org.springframework.security.oauth2.client.registration.ClientRegistration
import org.springframework.security.oauth2.core.AuthorizationGrantType
import org.springframework.security.oauth2.core.user.DefaultOAuth2User
import java.util.*
import java.util.Optional.empty
import java.util.Optional.of
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@ExtendWith(MockitoExtension::class)
class OAuth2SuccessHandlerTest {

    @Mock
    lateinit var userRepository: UserRepository

    @Mock
    lateinit var oAuth2AuthorizedClientService: OAuth2AuthorizedClientService

    @Mock
    lateinit var request: HttpServletRequest

    @Mock
    lateinit var response: HttpServletResponse

    @Mock
    lateinit var authentication: OAuth2AuthenticationToken

    lateinit var oAuth2SuccessHandler: OAuth2SuccessHandler

    lateinit var registryOAuth2Providers: MutableList<RegistryOAuth2Provider>

    @BeforeEach
    fun setup() {
        registryOAuth2Providers = ArrayList()
        oAuth2SuccessHandler = OAuth2SuccessHandler(userRepository, registryOAuth2Providers, oAuth2AuthorizedClientService)
        whenever(authentication.name).thenReturn("django")
    }

    @Test
    fun `onAuthenticationSuccess() should create user when it does not exist`() {
        // when
        whenever(userRepository.findById(anyString())).thenReturn(empty())
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication)

        // then
        val captor = forClass(User::class.java)
        verify(userRepository).save(captor.capture())
        assertThat(captor.value)
            .isNotNull
            .hasFieldOrPropertyWithValue("username", "django")
            .hasFieldOrPropertyWithValue("organization", null)
    }

    @Test
    fun `onAuthenticationSuccess() should update user when it exists`() {
        // given
        val user = User("calvin", null)

        // when
        whenever(userRepository.findById(anyString())).thenReturn(of(user))
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication)

        // then
        verify(userRepository).save(user)
    }

    @Test
    fun `onAuthenticationSuccess() should fill oAuth2User`() {
        // given
        val user = User("frankie", null)
        val oauth2User = OAuth2User("tarantino", "unchained", null)
        val client = mock(OAuth2AuthorizedClient::class.java)
        val registration = ClientRegistration
            .withRegistrationId("test_registration_id")
            .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
            .clientId("test_client_id")
            .redirectUriTemplate("test_uri_template")
            .authorizationUri("test_authorization_uri")
            .tokenUri("test_token_uri")
            .build()
        val principal = mock(DefaultOAuth2User::class.java)
        val oauth2Provider = mock(RegistryOAuth2Provider::class.java)
        registryOAuth2Providers.add(oauth2Provider)

        // when
        whenever(userRepository.findById(anyString())).thenReturn(of(user))
        whenever(oAuth2AuthorizedClientService.loadAuthorizedClient<OAuth2AuthorizedClient>(any(), anyString())).thenReturn(client)
        whenever(oauth2Provider.isAssignableFor(anyString())).thenReturn(true)
        whenever(oauth2Provider.getOAuth2User(principal, client)).thenReturn(oauth2User)
        whenever(client.clientRegistration).thenReturn(registration)
        whenever(authentication.principal).thenReturn(principal)
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication)

        // then
        assertThat(user).hasFieldOrPropertyWithValue("oAuth2User", oauth2User)
    }

    @Test
    fun `onAuthenticationSuccess() should redirect to home page`() {
        // when
        whenever(userRepository.findById(anyString())).thenReturn(empty())
        oAuth2SuccessHandler.onAuthenticationSuccess(request, response, authentication)

        // then
        verify(response).sendRedirect("/")
    }

}
